runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -10570,6 +10570,310 @@ inlineCompareAndSwapNative(
+   // We need to specify the encoding of this instruction so that the code generator knows we want the 16 byte version
+   TR::InstOpCode pmovmskb = TR::InstOpCode::PMOVMSKB4RegReg;
+   OMR::X86::Encoding pmovmskbEncoding = pmovmskb.getSIMDEncoding(&cg->comp()->target().cpu, TR::VectorLength128);
+
+
+   generateLabelInstruction(TR::InstOpCode::label, node, begLabel, cg);
+
+   // index = offset
+   generateRegRegInstruction(TR::InstOpCode::MOV4RegReg, node, indexReg, offsetReg, cg);
+
+   // limit = offset + length
+   generateRegMemInstruction(TR::InstOpCode::LEA4RegMem, node, limitReg, generateX86MemoryReference(offsetReg, lengthReg, 0, 0, cg), cg);
+
+   // loopLimit = (length & -16) + offset
+   generateRegRegInstruction(TR::InstOpCode::MOV4RegReg, node, loopLimitReg, lengthReg, cg);
+   generateRegImmInstruction(TR::InstOpCode::AND4RegImm4, node, loopLimitReg, -16, cg);
+   generateRegRegInstruction(TR::InstOpCode::ADD4RegReg, node, loopLimitReg, offsetReg, cg);
+
+   // If the 16 byte encoding of the pmovmskb instruction is not supported on this architecture,
+   // Prepare an 8 byte sign bit mask so we can run an alternate version of the algorithm
+   if (pmovmskbEncoding == OMR::X86::Encoding::Bad)
+      {
+      generateRegImm64Instruction(TR::InstOpCode::MOV8RegImm64, node, maskReg, 0x8080808080808080, cg);
+      }
+
+
+   generateLabelInstruction(TR::InstOpCode::label, node, loopLabel, cg);
+
+   // if index >= loopLimit, jump to handling the residual bytes
+   generateRegRegInstruction(TR::InstOpCode::CMP4RegReg, node, indexReg, loopLimitReg, cg);
+   generateLabelInstruction(TR::InstOpCode::JGE4, node, residualLabel, cg);
+
+
+   // If the 16 byte version of pmovmskb is supported on this architecture,
+   // we can proceed to generate code for the loop
+   if (pmovmskbEncoding != OMR::X86::Encoding::Bad)
+      {
+      // Load 16 bytes from address [buf + index]
+      generateRegMemInstruction(TR::InstOpCode::MOVDQURegMem, node, xmmChunkReg, generateX86MemoryReference(bufReg, indexReg, 0, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg), cg);
+
+      // Extract bitmask of sign bits
+      generateRegRegInstruction(TR::InstOpCode::PMOVMSKB4RegReg, node, maskReg, xmmChunkReg, cg, pmovmskbEncoding);
+
+      // Check if any negative values exist
+      generateRegRegInstruction(TR::InstOpCode::TEST2RegReg, node, maskReg, maskReg, cg);
+
+      // If the result is nonzero, we found at least one negative byte
+      generateLabelInstruction(TR::InstOpCode::JNE4,
+                               node,
+                               (recognizedMethod == TR::java_lang_StringCoding_hasNegatives) ? returnBooleanLabel : returnHasNegativesLabel,
+                               cg);
+
+      // increment index by 16 and jump back to the top of the loop
+      generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, indexReg, 16, cg);
+      generateLabelInstruction(TR::InstOpCode::JMP4, node, loopLabel, cg);
```

Comment:
These lines are the same in both the if and else branches.  I think they can be factored out of the if entirely.

Reply:
Good point! We increment the index by 16 in the vectorized case and by 8 in the non-vectorized case, but I can still pull them out and add a ternary operator to pick the amount we increment by.